### ISSUE TITLE
Error closing the session with .net 2.1.0-preview1

### ISSUE THREAD

**Issue by aspnet-hello** [aspnet/Home#2399]
_From @ssougnez on Tuesday, August 8, 2017 5:34:27 AM_

Hi,

As I have to use the SignalR for ASP.NET Core, I have to use the version 2.4.1-preview1 of .NET Core. My application requires the usage of sessions so I configured them this way (I removed some lines for clarity):

```
public void ConfigureServices(IServiceCollection services)
{
    services.AddDistributedSqlServerCache(options =>
    {
        options.ConnectionString = Configuration.GetConnectionString("DefaultConnection");
        options.SchemaName = "dbo";
        options.TableName = "Sessions";
    });

    services.AddSession(options =>
    {
        options.IdleTimeout = TimeSpan.FromDays(1);
        options.Cookie.Name = "RobotWarSessions";
    });
}

public void Configure(IApplicationBuilder app, IHostingEnvironment env, RobotWarContext dbContext)
{
    app.UseSession();
}
```

Then, I have an ASP.NET Core Web API route that just put information in the session:

`HttpContext.Session.SetString("data", data.AsJson());`

I also have a route that clear the session like this:

`HttpContext.Session.Clear();`

The whole thing works correctly and data are correctly set/removed from the session, however, the "dotnet run" command keeps on logging:
```
> fail: Microsoft.AspNetCore.Session.SessionMiddleware[1]
>       Error closing the session.
> System.OperationCanceledException: The operation was canceled.
>    at System.Threading.CancellationToken.ThrowOperationCanceledException()
>    at Microsoft.AspNetCore.Session.DistributedSession.<CommitAsync>d__32.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at Microsoft.AspNetCore.Session.SessionMiddleware.<Invoke>d__9.MoveNext()
```
This appears in the console every time I refresh the page. After some debugging, I noticed that it's coming from the `UseSession` line (go figure ^^), but I have no idea why it occurs...

Note that it only occurs when there are data in the session. Indeed, when I click "logout" in my application and refresh the page, I don't see the error. However, when I logged in (so put data in the session) and refresh the page, the error comes again.

A part of my code that I didn't show here is the fact that I also reading information in the session from a SIgnalR hub that stays open once the user is logged in

_Copied from original issue: aspnet/Session#184_

**Comment by aspnet-hello** [comment]
_From @Tratcher on Wednesday, August 9, 2017 9:07:08 AM_

2.4.1-preview1 of what package? That's not a version I recognize.

Does your SignalR hub stay active and read session beyond the lifetime of one request? ISession instances are considered per request resources and shouldn't be used outside of a request.

You get this error on page refresh for long running requests because the refresh aborts the current request. By default Session is persisted back to the cache only when a response completes. In this case the response failed so storing session will also fail (it uses the RequestAborted cancellation token). For long running requests you can preemptively commit your changes by calling ISession.CommitAsync
https://github.com/aspnet/HttpAbstractions/blob/5ec8a7134eaec5e878771e69b5b9b9c913ef9afe/src/Microsoft.AspNetCore.Http.Features/ISession.cs#L38

**Comment by aspnet-hello** [comment]
_From @ssougnez on Wednesday, August 9, 2017 9:34:00 AM_

Oops, sorry for the version, I meant to say "2.1.0-preview1-26462". There's so many version for everything that I got confused :-D

For your question, yes, my SignalR hub stays active and when it recieve data, it uses the session of the current context to retrieve data. Here is the main part of this:

```
public class Startup
{
    static internal IServiceProvider ServiceProvider { get; set; }

    public void ConfigureServices(IServiceCollection services)
    {
        services.AddSingleton<IHttpContextAccessor, HttpContextAccessor>();

        ServiceProvider = services.BuildServiceProvider();
    }
}
```

Then, the constructor of my hub:

```
public GameHub()
{
    if (Startup.ServiceProvider.GetService(typeof(IHttpContextAccessor)) is IHttpContextAccessor contextAccessor)
    {
        context = contextAccessor.HttpContext;
    }
}
```

And then, in a method of my Hub that is fired when data are sent through the socket, I do this:

`UserData data = context.Session.Get<UserData>("data");`

Basically, when the user logs in the application, he does it through a ASP.NET Core Web API route that creates data in the session, but then I need to be able to access the session from within my hub (in read/write) mode.

I'll try to use the `CommitAsync` to see if it solves the issue.

_To be honest, I'd like not to need a reference to the `HttpContext` in my hub but I don't see how else to do._

Thanks for the answer

**Comment by aspnet-hello** [comment]
_From @Tratcher on Wednesday, August 9, 2017 9:37:19 AM_

Session should NOT be used for authentication data.

You should ask SignlarR for recommendations on managing user state in hubs.

**Comment by aspnet-hello** [comment]
_From @davidfowl on Wednesday, August 9, 2017 9:38:22 AM_

That startup logic doesn't look good. You should not be building the service provider and storing it for later. You also shouldn't be setting the http context on the IHttpContextAccessor. Why are you doing any of that?

**Comment by aspnet-hello** [comment]
_From @ssougnez on Wednesday, August 9, 2017 9:44:07 AM_

I'm creating the service provider property to be able to retrieve a reference to services from within my Hub. Indeed, when the hub is constructed, it calls the parameterless constructor, therefore, I'm unable to use DI in the normal way. The `IHttpContextAccessor` was a way to retrieve a reference to the `HttpContext` => `ISession` from within my hub.

I agree that it's not the best solution but as I said, I need to retrieve a user Id stored in the session from within my hub and I don't know where else I can store it. Basically, when the user logs in, this creates a session, and as long as the session is valid, the user should be automatically logged in if he closes/refreshes the page.

**Comment by aspnet-hello** [comment]
_From @davidfowl on Wednesday, August 9, 2017 9:56:53 AM_

Storing a reference to the service provider defeated the entire purpose of inversion of control. I'm not due what version of SignalR you're using but hubs are activated by the DI system. What version are you using ? What's the name of the package?

There's an extension method to get the http context from the hub context. What does the session have to do with being logged in? What are you storing in the session?

**Comment by aspnet-hello** [comment]
_From @ssougnez on Wednesday, August 9, 2017 10:10:41 AM_

Hehe, I have the feeling that the architecture of my app is really bad (in a few post, I'll probably be ashamed ^^).

For the version that I'm using, it's the "1.0.0-alpha1-26462". I tried to inject the entity framework via the constructor but it didn't work, only the parameterless one was called and never the one with the database context (wasn't it the correct way to do so ?). For the extension method to access the `HttpContext`, I didn't know about that, which one is it?

Here's the complete idea behind my login process (try not to have a stroke at the end :-p).

- The user calls a ASP.NET Core Web API route that logs him in.
- If credentials are correct, the following information are stored in the session: UniqueId of the user in the database, the fact that the web socket connection is established or not and a token.

I'm using the token because I want to be able to recognize who is opening a connection to my hub. So after the data are stored in the session, the token (which is a simple GUID) is returned to the JavaScript app. 

Then:

- A web socket connection is opened
- Directly after, a specific message is sent through the socket connection with the token.
- The hub method is then checking if the token is the same as the one stored in session
- If it's not the same, the connection is closed
- If it's the same, the connection stays active, the token is removed from the session (this way, if a user gets this token in any way, he won't be able to use it) and the property that contains the fact that the socket connection is opened is set to true.

Now that I'm writing this explanation, I'm wondering if it wouldn't be easier to do the authentication also via the websocket, as my main issue is that I don't know how to be able to identify the previously logged in user once my web socket connection is established.

**Comment by aspnet-hello** [comment]
_From @davidfowl on Saturday, August 12, 2017 2:35:38 PM_

Cookie authentication is your friend here. Works well for websockets. Alternatively, you can use bearer tokens but you need to pass the token in a query string since the javascript websocket API doesn't support setting custom headers.

**Comment by aspnet-hello** [comment]
_From @azexpert on Tuesday, October 10, 2017 2:02:38 AM_

Hi team, 

I have the same issue with Session throwing error on page refresh:

```
info: Microsoft.AspNetCore.Sockets.Internal.Transports.WebSocketsTransport[1]                        
      10/10/2017 19:14:31: Connection Id 306434c8-1cbd-4f79-abff-23d8174b71d0: Socket closed.        
fail: Microsoft.AspNetCore.Session.SessionMiddleware[1]                                              
      Error closing the session.                                                                     
System.OperationCanceledException: The operation was canceled.                                       
   at System.Threading.CancellationToken.ThrowOperationCanceledException()                           
   at Microsoft.AspNetCore.Session.DistributedSession.<CommitAsync>d__32.MoveNext()                  
--- End of stack trace from previous location where exception was thrown ---                         
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()                                 
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)                      
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) 
   at Microsoft.AspNetCore.Session.SessionMiddleware.<Invoke>d__9.MoveNext()                             
```

I do also have core 2.1.0-preview with signalr:
```
    <PackageReference Include="Microsoft.AspNetCore.All" Version="2.1.0-preview1-27122" />
    <PackageReference Include="Microsoft.AspNetCore.SignalR" Version="1.0.0-alpha1-final" />                    
```

Sessions configuration is as per documentation:

```
            services.AddDistributedMemoryCache();

            services.AddSession(o =>
            {
                o.IdleTimeout = TimeSpan.FromDays(1);
                o.Cookie.HttpOnly = true;
                o.Cookie.Name = ".App.Session";
            });
```

I have nothing fancy in the Hub:

```
    public class DashboardHub : Hub<IDashboardClient>
    {
        private IHubUserTracker _userTracker;
        private IAppService _appService;
        private IMapper _mapper;

        public DashboardHub(IHubUserTracker userTracker, IAppService appService, IMapper mapper)
        {
            _userTracker = userTracker;
            _appService = appService;
            _mapper = mapper;
        }
```
Is there anything else that might be helpful? 
...
Please advise on how to gracefully close the session or at least how to fail quietly. 

Thanks!

**Comment by aspnet-hello** [comment]
_From @davidfowl on Tuesday, October 10, 2017 2:09:50 PM_

@azexpert can you make a project that reproduces the problem and put it on github with instructions?

**Comment by aspnet-hello** [comment]
_From @Tratcher on Tuesday, October 10, 2017 2:22:32 PM_

See https://github.com/aspnet/Session/issues/189

**Comment by aspnet-hello** [comment]
We periodically close 'discussion' issues that have not been updated in a long period of time.

We apologize if this causes any inconvenience. We ask that if you are still encountering an issue, please log a new issue with updated information and we will investigate.
